src/Symfony/Component/ObjectMapper/CachedObjectMapper.php

```diff
@@ -0,0 +1,351 @@
+        }
+
+        if (\is_string($target) && !class_exists($target)) {
+            throw new MappingException(\sprintf('Mapping target class "%s" does not exist for source "%s".', $target, get_debug_type($source)));
+        }
+
+        $cacheKey = str_replace(['\\', '/'], '_', $sourceClass.'-to-'.$targetClass);
+
+        if (!isset($this->cachedMappings[$cacheKey])) {
+            $cacheFile = $this->cacheDir.'/'.$cacheKey.'.php';
+
+            if (!file_exists($cacheFile) || $this->debug) {
+                $this->fs ??= new Filesystem();
+                $mappingData = $this->getMappingMetadata($source, $targetClass);
+                $code = $this->generateMappingCode($mappingData, $sourceClass, $targetClass);
+
+                $tmpFile = $this->fs->tempnam(\dirname($cacheFile), basename($cacheFile));
+
+                try {
+                    $this->fs->dumpFile($tmpFile, $code);
+                    $this->fs->rename($tmpFile, $cacheFile, true);
+                    $this->fs->chmod($cacheFile, 0o666 & ~umask());
+                } catch (IOException $e) {
+                    throw new MappingException(\sprintf('Failed to write "%s" mapping file.', $cacheFile), previous: $e);
+                }
+            }
+
+            $this->cachedMappings[$cacheKey] = require $cacheFile;
+        }
+
+        $mappingFunction = $this->cachedMappings[$cacheKey];
+        $mappingToObject = \is_object($target);
+        $targetObject = \is_object($target) ? $target : (new \ReflectionClass($targetClass))->newInstanceWithoutConstructor();
+
+        if ($source instanceof LazyObjectInterface) {
+            $source->initializeLazyObject();
+        } elseif (\PHP_VERSION_ID >= 80400) {
+            $refl = new \ReflectionClass($source);
+            if (!$refl->isUninitializedLazyObject($source)) {
+                $refl->initializeLazyObject($source);
+            }
+        }
+
+        $mappedTarget = $mappingFunction(
+            $source,
+            $targetObject,
+            $this->objectMapper ?? $this,
+            $this->metadataFactory,
+            $this->objectMap,
+            $this->propertyAccessor,
+            $this->transformCallableLocator,
+            $this->conditionCallableLocator,
+            $mappingToObject
+        );
+
+        if ($objectMapInitialized) {
+            $this->objectMap = null;
+        }
+
+        return $mappedTarget;
+    }
+
+    private function resolveTargetClass(object $source, object|string|null $target): ?string
+    {
+        if (\is_string($target)) {
+            return $target;
+        }
+
+        if (\is_object($target)) {
+            return $target::class;
+        }
+
+        // Try to find target class from metadata
+        $metadata = $this->metadataFactory->create($source);
+        if ($map = $this->getMapTarget($metadata, null, $source, null)) {
+            if ($map->target && \is_string($map->target) && class_exists($map->target)) {
```

Comment:
is the class_exists call legit? shouldn't `getMapTarget` not populate `$map->target` instead when this happens?